src/systemctl/systemctl-list-unit-files.c

```diff
@@ -3,33 +3,43 @@
 #include "ansi-color.h"
 #include "bus-error.h"
 #include "bus-locator.h"
+#include "path-util.h"
 #include "sort-util.h"
 #include "systemctl-list-unit-files.h"
 #include "systemctl-util.h"
 #include "systemctl.h"
 #include "terminal-util.h"
 
 static int compare_unit_file_list(const UnitFileList *a, const UnitFileList *b) {
-        const char *d1, *d2;
+        const char *d1, *d2, *a_id = NULL, *b_id = NULL;
+        int u, v;
 
         d1 = strrchr(a->path, '.');
         d2 = strrchr(b->path, '.');
 
         if (d1 && d2) {
-                int r;
-
-                r = strcasecmp(d1, d2);
+                int r = strcasecmp(d1, d2);
                 if (r != 0)
                         return r;
         }
 
-        return strcasecmp(basename(a->path), basename(b->path));
+        u = path_find_last_component(a->path, /* accept_dot_dot= */ false, /* next= */ NULL, &a_id);
+        v = path_find_last_component(b->path, /* accept_dot_dot= */ false, /* next= */ NULL, &b_id);
+
+        if (u < 0 || v < 0)  /* if we can't compare paths, compare error codes */
+                return CMP(u, v);
```

Comment:
so, i know i suggested to write it like this. But after https://github.com/systemd/systemd/pull/34391 I think we need to tweak this. i.e. we should never assume two strings are identical if they don't validate with the same error code. Hence tweak this further:

```c
if (u < 0 || v < 0) {
        r = CMP(u, v); /* Invalid paths? The compare by error code */
        if (r != 0)
                return r;
        return strcmp(a->path, b->path); /* Same error code? Then as last resort compare as strings */
}
```